dea_cogger/cogeo.py

```diff
@@ -264,7 +264,7 @@ def cog_translate(
                     for ij, w in wind:
                         matrix = src.read(window=w, indexes=indexes)
                         if nodata_mask is not None:
-                            matrix = numpy.array(matrix, dtype='int16')
+                            matrix = numpy.asarray(matrix, dtype='int16')
```

Comment:
If you want to minimize RAM usage, then should probably look into `np.copyto` with `where` parameter to avoid all the temporaries created by nodata erasing.

Reply:
Also why are we hard-coding `int16` data type?

Reply:
yeah not sure about the hard coded `int16` - maybe if there's no negative values its defaults to `uint8` or something? 🤷‍♂️

Reply:
also AFAIK `np.copyto` copies all the values from the source array to the destination causing the same memory duplication as `np.array(...)`

Reply:
and AFAIK, while something like:

```
arr = matrix[matrix == ​nodata_mask]
```

Calls `getitem` and copies data, 

```
matrix[matrix == ​nodata_mask] = nodata
```

calls `setitem` and updates `matrix` inplace.

Reply:
^ if that's what you were talking about @Kirill888 ?

Reply:
oh and I think we delete `matrix` after its written to the `MemoryFile`

```python
mem.write(matrix, ...)
del matrix
```

Reply:
Are these windows really that large, are you sure memory issue is with that? It's more likely due to whole output image is in memory uncompressed with all the overviews and all that. Probably setting some cheap compression for the first pass or backing it by disk would be more likely to help.

Reply:
`np.copyto` supports type conversion as well, but I see your point, you still have to create `where=(matrix==nodata_mask)` temporary array regardless, so probably no savings there.

Reply:
tbh... I don't have any memory issues I was just browsing the code and felt like optimising so maybe this is completely uneccessary. :)

Reply:
actually no, copyto won't help at all here.

but can you explain how new code is different from the old? One has to make a copy of `matrix` regardless, so that nodata values can be remapped without affecting source data.

Reply:
Oh woops I didn't realise the source data had to preserved, in that case my change doesn't work. My change was just edditing the underlying array of `matrix` inplace. I mistakenly thought that the originals weren't being written to the GeoTIFFs

Reply:
sorry this is using rasterio not netcdf/h5py, so it's already a copy. Still I suggest we leave it without changes if it's not causing issues, I would check with Damien. Also I usually use `matrix.astype(..)` syntax instead.